core/src/main/scala/kafka/log/ProducerIdMapping.scala

```diff
@@ -0,0 +1,387 @@
+          .set(TimestampField, entry.timestamp)
+        pidEntryStruct
+    }.toArray
+    struct.set(PidEntriesField, entriesArray)
+
+    val buffer = ByteBuffer.allocate(struct.sizeOf)
+    struct.writeTo(buffer)
+    buffer.flip()
+
+    // now fill in the CRC
+    val crc = Crc32.crc32(buffer, PidEntriesOffset, buffer.limit - PidEntriesOffset)
+    ByteUtils.writeUnsignedInt(buffer, CrcOffset, crc)
+
+    val fos = new FileOutputStream(file)
+    try {
+      fos.write(buffer.array, buffer.arrayOffset, buffer.limit)
+    } finally {
+      fos.close()
+    }
+  }
+
+  private def verifyFileName(name: String): Boolean = FilenamePattern.findFirstIn(name).isDefined
+
+  private def offsetFromFile(file: File): Long = {
+    s"${file.getName.replace(s".$FilenameSuffix", "")}".toLong
+  }
+
+  private def formatDirName(topicPartition: TopicPartition): String = {
+    s"$DirnamePrefix-${topicPartition.topic}-${topicPartition.partition}"
+  }
+
+  private def formatFileName(lastOffset: Long): String = {
+    s"$lastOffset.$FilenameSuffix"
+  }
+
+}
+
+/**
+ * Maintains a mapping from ProducerIds (PIDs) to metadata about the last appended entries (e.g.
+ * epoch, sequence number, last offset, etc.)
+ *
+ * The sequence number is the last number successfully appended to the partition for the given identifier.
+ * The epoch is used for fencing against zombie writers. The offset is the one of the last successful message
+ * appended to the partition.
+ */
+@nonthreadsafe
+class ProducerIdMapping(val config: LogConfig,
+                        val topicPartition: TopicPartition,
+                        val snapParentDir: File,
+                        val maxPidExpirationMs: Int) extends Logging {
+  import ProducerIdMapping._
+
+  val snapDir: File = new File(snapParentDir, formatDirName(topicPartition))
+  snapDir.mkdir()
+
+  private val pidMap = mutable.Map[Long, ProducerIdEntry]()
+  private var lastMapOffset = -1L
+  private var lastSnapOffset = -1L
+
+  /**
+   * Returns the last offset of this map
+   */
+  def mapEndOffset = lastMapOffset
+
+  /**
+   * Get a copy of the active producers
+   */
+  def activePids: immutable.Map[Long, ProducerIdEntry] = pidMap.toMap
+
+  /**
+   * Load a snapshot of the id mapping or return empty maps
+   * in the case the snapshot doesn't exist (first time).
+   */
+  private def loadFromSnapshot(logEndOffset: Long) {
+    pidMap.clear()
+    lastSnapshotFile(logEndOffset) match {
+      case Some(file) =>
+        try {
+          trace("Loading snapshot from file: %s".format(file.toPath))
+          loadSnapshot(file, pidMap)
+          lastSnapOffset = offsetFromFile(file)
+          lastMapOffset = lastSnapOffset
+        } catch {
+          case e: CorruptSnapshotException =>
+            // TODO: Delete the file and try to use the next snapshot
+            throw e
+        }
+
+      case None =>
+        lastSnapOffset = -1
+        lastMapOffset = -1
+        snapDir.mkdir()
```

Comment:
Actually, I undeleted these lines as they are needed. The `loadSnapshot` can be called in two cases: During initial start, and also when the log is truncated. In the latter case, if there is no previous snapshot, we need to reset to the start offset, which is what this code does.